src/netbase.cpp

```diff
@@ -48,38 +48,57 @@ std::vector<CNetAddr> WrappedGetAddrInfo(const std::string& name, bool allow_loo
     // return addresses whose family we have an address configured for.
     //
     // If we don't allow lookups, then use the AI_NUMERICHOST flag for
     // getaddrinfo to only decode numerical network addresses and suppress
     // hostname lookups.
-    ai_hint.ai_flags = allow_lookup ? AI_ADDRCONFIG : AI_NUMERICHOST;
-
-    addrinfo* ai_res{nullptr};
-    const int n_err{getaddrinfo(name.c_str(), nullptr, &ai_hint, &ai_res)};
-    if (n_err != 0) {
-        return {};
-    }
-
-    // Traverse the linked list starting with ai_trav.
-    addrinfo* ai_trav{ai_res};
-    std::vector<CNetAddr> resolved_addresses;
-    while (ai_trav != nullptr) {
-        if (ai_trav->ai_family == AF_INET) {
-            assert(ai_trav->ai_addrlen >= sizeof(sockaddr_in));
-            resolved_addresses.emplace_back(reinterpret_cast<sockaddr_in*>(ai_trav->ai_addr)->sin_addr);
+    const int ai_flags = allow_lookup ? AI_ADDRCONFIG : AI_NUMERICHOST;
+
+    // getaddrinfo is called twice and to prevent duplicates a set is used
+    // std::set used over a std::unordered_set as CNetAddr does not have a hash function
+    std::set<CNetAddr> resolved_addresses_set;
+
+    auto callgetaddrinfo = [&](int flags, bool only_add_local_addr) {
+        ai_hint.ai_flags = flags;
+        addrinfo* ai_res{nullptr};
+        const int n_err{getaddrinfo(name.c_str(), nullptr, &ai_hint, &ai_res)};
+        if (n_err != 0) {
+            return;
         }
-        if (ai_trav->ai_family == AF_INET6) {
-            assert(ai_trav->ai_addrlen >= sizeof(sockaddr_in6));
-            const sockaddr_in6* s6{reinterpret_cast<sockaddr_in6*>(ai_trav->ai_addr)};
-            resolved_addresses.emplace_back(s6->sin6_addr, s6->sin6_scope_id);
+
+        // Traverse the linked list of addrinfo structures
+        addrinfo* ai_trav{ai_res};
+        while (ai_trav != nullptr) {
+            CNetAddr addr;
+            if (ai_trav->ai_family == AF_INET) {
+                assert(ai_trav->ai_addrlen >= sizeof(sockaddr_in));
+                addr = CNetAddr{reinterpret_cast<sockaddr_in*>(ai_trav->ai_addr)->sin_addr};
+            }
+            if (ai_trav->ai_family == AF_INET6) {
+                assert(ai_trav->ai_addrlen >= sizeof(sockaddr_in6));
+                const sockaddr_in6* s6{reinterpret_cast<sockaddr_in6*>(ai_trav->ai_addr)};
+                addr = CNetAddr{s6->sin6_addr, s6->sin6_scope_id};
+            }
+            if (!only_add_local_addr || addr.IsLocal()) {
+                resolved_addresses_set.emplace(addr);
+            }
+            ai_trav = ai_trav->ai_next;
         }
-        ai_trav = ai_trav->ai_next;
+        freeaddrinfo(ai_res);
+    };
+
+    callgetaddrinfo(ai_flags, false);
+
+    // AI_ADDRCONFIG on some systems may exclude loopback only addresses, especially IPV6
+    // We perform a second lookup with ai_flags set to 0 and add local addresses to the set
+    if (ai_flags == AI_ADDRCONFIG) {
+        callgetaddrinfo(0, true);
```

Comment:
Instead of using 0 as the first argument, might be better to remove only `AI_ADDRCONFIG`, e.g. something like
```c++
callgetaddrinfo(ai_flags & (~AI_ADDRCONFIG), true);
```

Reply:
`ai_flags` only had `AI_ADDRCONFIG` set so without it it would be 0 anyway right?

Reply:
Yeah, it would.  The thought here is to increase resilience in the code, so that if ai_flags changes we're not hard-coded to 0.  Seems like the scope of this `if` condition is intended as "if AI_ADDRCONFIG set" rather than "is only AI_ADDRCONFIG set" (which is now updated in f95fb4a1d826765553f3319ee6f1024094954073).  I'm thinking we would want `callgetaddrinfo()` to use similar logic (i.e. keep flags defined except for AI_ADDRCONFIG).

Reply:
Good suggestion thanks, implemented.